### PR TITLE
Fix stroke merging logic bugs

### LINKED ISSUE #1217: Strokes are merged in incorrect order in some cases
While creating visualizations for the Milestone 6/final implementation report, I discovered a few subtle bugs in the stroke merging logic:

1. Identical consecutive control points are included in the final merged strokes. I suspect they might screw up or bias the B-spline simplification logic, so because of that should be left out.
2. The orientation preservation logic is slightly wrong. It breaks in cases like: `-->  <--  <--  -->` (processing from left to right). In this case we should reverse the middle two strokes (or the outer two strokes) but what actually happens at the moment is that we flip the second stroke and the final stroke, getting `--> --> <-- <--`. The fix is easy to code, though slightly hard to follow: We should reverse the next stroke if and only if the next stroke is oriented consistent with the current stroke and the current stroke is going to be reversed.
3. Sometimes the clustered strokes are concatenated in the reverse of the intended control point order. This results in some messed-up looking strokes. We have enough of those already without merging, so I want to avoid that. Fortunately, the fix is straightforward: Just reverse the list of strokes if, after reorienting strokes, the second stroke doesn't begin where the first one ends.

I've implemented these fixes and will get a PR up shortly.

### PR DESCRIPTION
**Pull request by spigo900**
Closes #1217.

### CODE REVIEW

**Review by lichtefeld — CHANGES_REQUESTED** [review]

File: adam_preprocessing/singlefile_stroke_extraction.py [hunk 1]
```diff
@@ -32,13 +32,26 @@ def main():
...
+        "--merge-small-strokes",
+        action="store_true",
+        help="Merge small strokes before filtering."
+    )
+    parser.add_argument(
+        "--no-merge-small-strokes",
+        action="store_false",
+        dest="merge_small_strokes",
+        help="Don't merge small strokes before filtering."
+    )
```
[the remark this answers is not in the document]
Comment (spigo900) [reply adam_preprocessing/singlefile_stroke_extraction.py:45]: It's so that you can disable the behavior when using a bash script that passes `--merge-small-strokes`. This lets you pass `--no-merge-small-strokes` via the `"$@"` args. In this case I don't expect we'll need to, but I've run into this case of wanting to override the defaults before with scripts that pass lots of boolean defaults, specifically Hugging Face model training/inference scripts. With the Hugging Face argument parser you can pass an explicit boolean value `--merge-small-strokes True` or `--merge-small-strokes False` but argparse doesn't support that i don't think.

For example, say you are invoking stroke extraction from a bash script. You pass `--merge-small-strokes` as part of the bash script. You also pass any leftover bash args to the Python script. If you want to invoke the bash script but disable merging for whatever purpose, the `store_true` alone isn't enough -- you'd have to either edit it out (works fine for debugging, but not as much for other things) or have a `store_false` flag you can pass to disable the behavior.

I'm fine with deleting this, though, because it doesn't seem necessary in this script.

[the remark this answers is not in the document]
Comment (lichtefeld) [reply adam_preprocessing/singlefile_stroke_extraction.py:45]: Ah makes sense. I've just not encountered a need to do such a thing :) -- happy to leave this in. I almost just approved and asked the question at the same time.

**Review by lichtefeld — APPROVED** [review]
